ompi/runtime/ompi_mpi_init.c

```diff
@@ -268,6 +269,59 @@ MPI_Fint *MPI_F08_STATUSES_IGNORE = NULL;
 
 #include "mpif-c-constants.h"
 
+static const char *ompi_thread_level_keywords[] = {"single", "serialized", "funneled", "multiple"};
+static const char *ompi_thread_level_prepositions[] = {"mpi_thread_", "thread_", NULL};
+/* In the future integer MPI_ABI values for MPI_THREAD_SINGLE-MULTIPLE may be
+ * non-sequential (but ordered) integer values. If you are implementing MPI
+ * ABI changes please refer to
+ * https://github.com/open-mpi/ompi/pull/13211#discussion_r2085086844
+ */
+static const int ompi_thread_level_values[] = {MPI_THREAD_SINGLE, MPI_THREAD_SERIALIZED,
+                                               MPI_THREAD_FUNNELED, MPI_THREAD_MULTIPLE};
+
+int ompi_getenv_mpi_thread_level(int *requested)
+{
+    char* env;
+    if (NULL != (env = getenv("OMPI_MPI_THREAD_LEVEL"))) {
+        char *prep = NULL, *token = (char *) env /* full match */;
+        int pidx = 0, found = strtol(env, &prep, 10);
+
+        if (prep == env) {  /* no digits found */
+            found = -1;
+            while (NULL != (prep = (char *) ompi_thread_level_prepositions[pidx])) {
+                if (0 == strncasecmp(prep, env, strlen(prep))) {
+                    token = env + strlen(prep);
+                    break; /* got a token let's find a match */
+                }
+                pidx++;
+            }
+            const int nb_keywords = sizeof(ompi_thread_level_keywords)/sizeof(ompi_thread_level_keywords[0]);
+            for (int i = 0; i < nb_keywords; i++) {
+                if (0 == strncasecmp(ompi_thread_level_keywords[i], token, strlen(token))) {
+                    if (-1 != found) { /* not the first match, bail out */
+                        return OMPI_ERR_BAD_PARAM;
+                    }
+                    found = i;
+                }
+            }
+        } else {  /* at least a digit has been found */
+            if (*prep != '\0') {  /* non-numerical left-over */
+                return OMPI_ERR_BAD_PARAM;
+            }
+            /* did we found a valid integer value? */
+            const int nb_keywords = sizeof(ompi_thread_level_values)/sizeof(ompi_thread_level_values[0]);
+            if (found >= nb_keywords) {
+                return OMPI_ERR_BAD_PARAM;
+            }
```

Comment:
I don't quite grok this -- this works for the integer values 0, 1, 2, 3, but won't work if -- when ABI is merged in -- the user specifies the ABI integer value for `MPI_THREAD_MULTIPLE` (4096), for example.

Wouldn't it be better to iterate through the values of the ompi_thread_level_values and see if the integer `found` value matches one of them?

Reply:
The documentation states 0 to 3, the MCA help string states 0 to 3, the previous OMPI versions were also based on 0 to 3 values. If we add the fact that the internal values, ABI or not, are hard to know for users and are not necessarily monotonically increasing, I don't think that accepting the internal values makes sense or is usefull for users (except maybe nitpicking).

Reply:
We've previously allowed 0-3, which is probably "well known" (or at least easy to guess).

The ABI values are published in the MPI-5 document, so those fit at least some definition of "well known", too.

Meaning: if we're going to accept integer values, we should probably accept *both* 0-3 and the ABI values that are published in the MPI spec.

Reply:
I'm not saying that they *have* to.  I'm saying that if we offer the ability to use integer values, we should offer *both* sets of integer values because *someone* will try to set the ABI value and be surprised that it didn't work.

Reply:
I don't have a strong opinion on the ABI values, 

but I will need this in v5, where there is no ABI support planned anyway
> Setting OMPI_MPI_THREAD_LEVEL to a value different than requested in MPI_Init_thread would invoke the error handler, even though it is an useful override in some threaded library use cases.

I would recommend deferring to a separate PR when we implement ABI for v6, if only to keep the cherry-picking easier.